Create DiagonalCoulombHamiltonian class

kevinsung opened the pull request:
Includes conversion routines to/from FermionOperator and an optimized Jordan-Wigner Transform.

I'm going to say this resolves #246 ; if we want more features we can create a new issue.

kevinsung commented:
Reverting the changes made in #305 (seems likely to happen) will resolve the conflicts.

babbush commented:
please resolve conflicts. @kevinsung your last few PRs have been out of sync with the repo.

kevinsung commented:
The conflict is only due to the change made in #305 so if we revert that first then this will go through.

babbush commented on src/openfermion/transforms/_jordan_wigner.py:
```diff
@@ -61,9 +65,45 @@ def jordan_wigner(operator):
     return transformed_operator
 
 
+def jordan_wigner_diagonal_coulomb_hamiltonian(operator):
```
Perhaps this should be called by the standard jordan_wigner function when it is passed a DiagonalCoulombHamiltonian. That way one does not need to keep track of multiple different jordan_wigner functions.

kevinsung replied:
I did do it that way. However, this reminds me that I should also update `get_sparse_operator`.

babbush commented on src/openfermion/ops/__init__.py:
```diff
@@ -11,6 +11,7 @@
 #   limitations under the License.
 
 from ._binary_polynomial import BinaryPolynomial
+from ._diagonal_coulomb_hamiltonian import DiagonalCoulombHamiltonian
```
I am not sure I like "DiagonalCoulombHamiltonian" because it seems to exclude other operators that might not have any explicit connection to Coulomb operators or Hamiltonians, which also fit this form. Sorry if I agreed to this name in the past...

what about DiagonalTwoBodyOperator?

babbush replied:
Maybe this name is okay. If you like, let's just keep it.

kevinsung replied:
I had the same thought, but I think DiagonalTwoBodyOperator sounds too much like it doesn't contain one-body terms.

babbush commented on src/openfermion/ops/_diagonal_coulomb_hamiltonian.py:
```diff
@@ -0,0 +1,39 @@
+        one_body(ndarray): The Hermitian matrix :math:`T`.
+        two_body(ndarray): The real symmetric matrix :math:`V`.
+        constant(float): The constant.
+        n_qubits(int): The number of qubits.
```
why does n_qubits need to be an attribute? I don't necessarily like that because our other classes like FermionOperator and QubitOperator don't have this attribute and n_qubits is easily computed from the dimension of one_body or two_body.

kevinsung replied:
I only did it because PolynomialTensor has this attribute and this class is more similar to that than to FermionOperator and QubitOperator. Do you think we should remove it from PolynomialTensor too?

kevinsung replied:
Ok I removed this attribute from DiagonalCoulombHamiltonian.

kevinsung commented:
Ok I've addressed these comments and also slightly modified the conversion routine from FermionOperator to check that the two-body coefficients are real and discard small imaginary parts.
